Atomic.md

```diff
@@ -0,0 +1,270 @@
+         not empty.
+
+    -   This folder will never be enumerated as the kext will exit when
+         we do not see the InVirtualizationRoot bit
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Folder is marked as in virtualization root but not empty.
+
+        -   This folder will never be enumerated as the kext will exit
+             when we do not see the IsEmpty bit
+
+    -   Folder is marked as empty but not in virtualization root
+
+        -   This folder will never be enumerated as the kext will exit
+             when we do not see the InVirtualizationRoot bit (we'll
+             never reach the IsEmpty test)
+
+WritePlaceholderFile:
+
+-   Fopen() fails to create a file
+
+    -   This should be retriable as a second attempt should see the
+         IsEmpty bit on the parent folder and call WritePlaceholderFile
+         again. No partial placeholders were left in the parent
+         directory.
+
+-   Ftruncate fails to expand the file to the correct size (or is never
+     reached because of a crash)
+
+    -   We have a partial placeholder file in the virtualization root
+         that cannot be recovered. Future attempts of
+         WritePlaceholderFile will fail because fopen(path, "wbx")
+         won't let us overwrite the existing file.
+
+-   Memcpy'ing the providerID xattr fails (or is never reached because
+     of a crash)
+
+    -   We have a partial placeholder file in the virtualization root
+        that cannot be recovered. There are no flags set on this file.
+
+    -   The MirrorProvider doesn't currently use the providerID for
+         anything (although it should fail if its empty)
+
+    -   VFSForGit will be unable to parse the placeholder version from
+         the providerID, OnGetFileStream will fail so hydrating this
+         file will fail. Retrying will not help.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have a partial placeholder file in the virtualization root
+        that cannot be recovered. There are no flags set on this file.
+
+
+    -   The MirrorProvider doesn't use contentID so no issues will occur
+
+    -   VFSForGit will be unable to get the SHA of the object that it
+         needs to download and hydrating this file will fail. Retrying
+         will not help.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   Placeholder appears to be not in virtualization root and not
+         empty.
+
+    -   This file will never be enumerated as the kext will exit when we
+         do not see the InVirtualizationRoot bit
+
+-   One of the SetBitInFileFlags calls fails in
+     InitializeEmptyPlaceholder
+
+    -   Placeholder is marked as in virtualization root but not empty.
+
+        -   This file will never be hydrated as the kext will exit when
+             we do not see the IsEmpty bit
+
+    -   Placeholder is marked as empty but not in virtualization root
```

Comment:
How can this happen?  If VFSForGit fails to mark the placeholder as inside the root does it continue onto the next (is empty) bit?

Reply:
It should not be possible, but I wanted to call it out either way. Perhaps it's clutter and should be cut then?

Reply:
> It should not be possible, but I wanted to call it out either way. Perhaps it's clutter and should be cut then?

My vote is to cut it.  There's already a lot of failure modes to take in, and it might make it easier to reason about the best fix if we don't also consider the impossible scenarios.

(I think it *is* possible for a file to end up like this from a user mucking with it, but that's unrelated to the focus of your document)

Reply:
That makes sense, will do.